### ISSUE TITLE
3.3.1 irepl crossed the streams (wrong data written, SYS_COPY_LEN_ERR)

### ISSUE THREAD

**Issue by mcast** [irods/irods#3319]
Content intended for one file was written to another file during part of several concurrent `irepl -N2`incantations, before all files failed with `SYS_COPY_LEN_ERR`.

This was on irods-3.3.1 (server and standard icommands), so I'm not expecting it to be fixed, but it seems serious enough to let other users know it can happen and maybe go looking for the cause or an equivalent in 4.1.x.
## Errors seen

These were the times and errors, all on 2016-10-03.  Two pairs of `irepl` failing in the same second.  Incomplete files are left behind on the `unix file system` vault.

| start | end | error |
| --- | --- | --- |
| 05:39:34 | 05:39:41 | ERROR: replUtil: repl error for /cgp/unaligned_import/study/6432/151548.bam, status = -27000 status = -27000 SYS_COPY_LEN_ERR |
| 05:39:34 | 05:39:41 | ERROR: replUtil: repl error for /cgp/unaligned_import/study/6432/151516.bam, status = -27000 status = -27000 SYS_COPY_LEN_ERR |
| 10:08:26 | 10:08:29 | ERROR: replUtil: repl error for /cgp/unaligned_import/study/6488/167891.bam, status = -27000 status = -27000 SYS_COPY_LEN_ERR |
| 10:08:24 | 10:08:29 | ERROR: replUtil: repl error for /cgp/unaligned_import/study/6432/128165.bam, status = -27000 status = -27000 SYS_COPY_LEN_ERR |

Repeating the `irepl` commands produced correct copies, md5sum verified.  The old incomplete files are renamed to `/irods-cgp-i01-fg/orphan/unaligned_import/...` in this process.

So far, this just looks like transient failure and is fixed by repeating.
## Data crossover seen

In short, when replicating both files (A, B) concurrently, pieces of source file A may be found either in destination file A or destination B.  There should never be pieces of file A in file B, they are different files.

What drew my attention was that the sizes of orphan/.../1515**16**.bam.1002011531 and .../1515**48**.bam were the same, to the byte, at about 363 MiB.  So I chopped the files into 64KiB pieces and [looked for duplicate sha1sums](https://github.com/mcast/scripts/commits/master/bin/fragsum) across all eight files.

| file | range/64KiB | state |
| --- | --- | --- |
| .../151548.bam | 0..5813 | valid data |
| .../151516.bam | 0..4763 | valid data |
| .../167891.bam | 0..2546 | valid data |
| .../128165.bam | 0..3516 | valid data |
| orphan/.../151548.bam.741175103 | 0..2905 | valid data |
| orphan/.../151548.bam.741175103 | 2906 | unique part block |
| orphan/.../151516.bam.1002011531 | 0..2905 | nul |
| orphan/.../151516.bam.1002011531 | 2906 | unique block |
| orphan/.../151**516**.bam.1002011531 | 2907..5813 | match 151**548**.bam 2907..5813 |
| orphan/.../167891.bam.134887095 | 0..1272 | nul |
| orphan/.../167891.bam.134887095 | 1273 | unique block |
| orphan/.../167891.bam.134887095 | 1274..2546 | valid data |
| orphan/.../128**165**.bam.1335084847 | 0..1272 | match 167**891**.bam 0..1272 |
| orphan/.../128165.bam.1335084847 | 1273 | unique part block |

Unique (part) blocks are invalid data, I presume they are a mixture of the two source and/or nuls.  They are unique in that the sha1(64k block) was not seen already in the valid data.

Exact byte sizes and file times

```
 381009900 Oct  3 05:39 /irods-cgp-i01-fg/orphan/unaligned_import/study/6432/151516.bam.1002011531
 381009900 Oct  4 05:03 /irods-cgp-i01-fg/unaligned_import/study/6432/151548.bam
 312180020 Oct  4 05:29 /irods-cgp-i01-fg/unaligned_import/study/6432/151516.bam
 230425971 Oct  4 05:21 /irods-cgp-i01-fg/unaligned_import/study/6432/128165.bam
 190504950 Oct  3 05:39 /irods-cgp-i01-fg/orphan/unaligned_import/study/6432/151548.bam.741175103
 166897844 Oct  3 10:08 /irods-cgp-i01-fg/orphan/unaligned_import/study/6488/167891.bam.134887095
 166897844 Oct  3 20:20 /irods-cgp-i01-fg/unaligned_import/study/6488/167891.bam
  83448922 Oct  3 10:08 /irods-cgp-i01-fg/orphan/unaligned_import/study/6432/128165.bam.1335084847
```
## More?

During replication of ~800 TiB in ~2e6 files, I have seen (I guess) a few dozen SYS_COPY_LEN_ERR errors and still don't know what provokes them.

The files are probably still in the vaults but I haven't looked at them with the (grotty hack) fragsum tool, or anything else that would spot mismatched file sizes or (like a minhash) wrong-source contents.

@kript (local sysad) - [internal link](https://gitlab.internal.sanger.ac.uk/CancerIT/cgpDataOut/issues/29#note_3083) has more detail.

**Comment by trel** [comment]
@mcast apologies for the radio silence.

have you seen this behavior since oct 2016?

i feel comfortable closing this if you have not seen anything similar in 4.1.x.

**Comment by mcast** [comment]
No, I haven't seen it since.  We are still on 3.3.1.

Also I have not run such heavy loads like that since: IIRC there were around 12 * `irepl -N2`s, from four ires over a 5*2Gbit/sec link onto two ires, each host with three partitions of RAID.

It looks like a bug which submerged itself back below the waves...

**Comment by trel** [comment]
Okay, I'm going to close it - there have been a great number of things fixed since 3.3.1.

Thanks.
